### PR TITLE
Improved Authentication Logic

### PR DESCRIPTION
**Pull request by RobertWetzler**
Big PR, here's what's mainly changed:

Frontend:
- Navigation gaurds: checks if the user is logged in before allowing them to load pages that require authentication. The authenticated routes are defined in /crosswars-frontend/src/routes.js and the navigation logic is in index.js in the same directory.
- Navigation gaurds will, if user not authenticated/has expired JWT, take the user to the login page, which will then afterwards direct them to the page they were trying to get to previously (by passing nextUrl router query)
- Improved login button to look like a legit login with google button
- Login button checks if user exists and if not makes a POST request

Backend: 
- Auth'd GET user endpoints now don't require passing in a user_id. If user_id not passed, it will return the User of the JWT token. This is useful for getting the current user on the frontend upon login.
- POST a new user now uses a a different Authenticated annotation, because the original one was trying to map the JWT to a user in the database (which will never work for creating new users). Instead, the annotation @AuthenticatedNewUser just makes auth_user a new User  using the JWT data, and then stores in in the database

### CODE REVIEW

File: crosswars-backend/src/main/kotlin/xyz/crosswars/service/GoogleAuthService.kt [hunk 1]
```diff
@@ -43,7 +43,16 @@ class GoogleAuthService(
...
         } ?: throw UnauthorizedException("Google id token is invalid")
+
+    fun createUserFromToken(idToken: String): User =
+            // TODO: see if there's a way to handle token expired exceptions
+            verifier.verify(idToken)?.let { googleIdToken ->
+                val payload: GoogleIdToken.Payload = googleIdToken.payload
+                val userId: String = payload.subject
+                val email: String = payload.email
+                val givenName: String = payload["given_name"] as String
+                User(userId = userId, name=givenName, email=email, remind=false)
```
Comment (loomisdf) [inline crosswars-backend/src/main/kotlin/xyz/crosswars/service/GoogleAuthService.kt:56]: why not simplify this section and remove the single use variables
```Kotlin
val payload: GoogleIdToken.Payload = googleIdToken.payload
User(
    userId = payload.subject, 
    name = payload["given_name"] as String, 
    email = payload.email, 
    remind = false
)
```

File: crosswars-frontend/src/pages/GroupPage.vue [hunk 2]
```diff
@@ -102,7 +102,7 @@ export default defineComponent({
   },
   computed: {
     inviteLink: function(): string {
-        return `${location.host}/#/group_invite/${this.group.id}`
+        return `${location.host}/#/groupInvite/${this.group.id}`
```
Comment (loomisdf) [inline crosswars-frontend/src/pages/GroupPage.vue:105]: Did it not like the underscores?

**Review by loomisdf — CHANGES_REQUESTED** [review]

**Comment by loomisdf** [comment]
Other than the one suggestion I made, everything looks good.

**Review by loomisdf — APPROVED** [review]

**Comment by RobertWetzler** [comment]
Authorized getGroupsByUser, otherwise it would be pretty easy for someone to fingerprint every group a given user is in, and then construct group invite links for each group. As a convention, retrieving a given group's ID should require authorization, so you can only get IDs of the groups you belong to.

**Comment by RobertWetzler** [comment]
Also with a previous commit - group IDs are now generated randomly, instead of being derived from the group name. That way you can't construct a group invite link just from knowing a group's name

**Comment by RobertWetzler** [comment]
Also with other previous commit - limited users to only creating 5 groups, otherwise user's could spam group creations and overwhelm the backend

**Pull request merged**